### PR TITLE
Translate new set of Intel FPGA Loop Controls

### PR DESCRIPTION
**Pull request by vmaksimo**
Specification can be found here:
https://github.com/KhronosGroup/SPIRV-Registry/pull/62

As per revision C -> revision E transition in the spec, patch introduces
translation of the following loop controls:
* PipelineDisableINTEL
* LoopCoalesceINTEL
* MaxInterleavingINTEL
* SpeculatedIterationsINTEL

### CODE REVIEW

File: lib/SPIRV/SPIRVReader.cpp [hunk 1]
```diff
@@ -784,6 +784,47 @@ void SPIRVToLLVM::setLLVMLoopMetadata(const LoopInstType *LM,
...
+    Metadata.push_back(llvm::MDNode::get(
+        *Context,
+        getMetadataFromNameAndParameter("llvm.loop.intel.pipelining.enable",
+                                        LoopControlParameters[NumParam])));
+    ++NumParam;
+    assert(NumParam <= LoopControlParameters.size() &&
+           "Missing loop control parameter!");
+  }
+  if (LC & LoopControlLoopCoalesceINTEL) {
+    if (LoopControlParameters.size()) {
```
Comment (mlychkov) [inline lib/SPIRV/SPIRVReader.cpp]: Should we check params size with NumParam value?

File: test/transcoding/FPGALoopAttr.ll [hunk 2]
```diff
@@ -168,6 +168,175 @@ for.end36:                                        ; preds = %for.cond29
   ret void
 }
 
+; Function Attrs: noinline nounwind optnone
```
Comment (AGindinson) [inline test/transcoding/FPGALoopAttr.ll]: This test and `FPGALoopMergeInst.ll` are practically identical. I'd suggest moving these new checks to the better-documented `FPGALoopMergeInst.ll` (adding the source code to the comment section as well).

File: lib/SPIRV/SPIRVReader.cpp [hunk 3]
```diff
@@ -784,6 +784,47 @@ void SPIRVToLLVM::setLLVMLoopMetadata(const LoopInstType *LM,
       Metadata.push_back(llvm::MDNode::get(*Context, Parameters));
     }
   }
+  if (LC & LoopControlPipelineEnableINTEL) {
+    Metadata.push_back(llvm::MDNode::get(
+        *Context,
+        getMetadataFromNameAndParameter("llvm.loop.intel.pipelining.enable",
+                                        LoopControlParameters[NumParam])));
+    ++NumParam;
```
Comment (AGindinson) [inline lib/SPIRV/SPIRVReader.cpp]:
```suggestion
                                        LoopControlParameters[NumParam++])));
```
..., which could apply to several sections below.

File: test/transcoding/FPGALoopAttr.ll (hunk 2 above)
Comment (vmaksimo) [reply test/transcoding/FPGALoopAttr.ll]: Are you suggesting only moving new checks or duplicating is also fine (although it's obviously not)?

Comment (AGindinson) [reply test/transcoding/FPGALoopAttr.ll]: My suggestion is to move the new checks; and to remove this file altogether (the latter probably deserves a separate PR so as not to confuse ourselves & the project maintainers).

File: lib/SPIRV/SPIRVReader.cpp (hunk 1 above)
Comment (vmaksimo) [reply lib/SPIRV/SPIRVReader.cpp]: No, these two are not related

File: lib/SPIRV/SPIRVReader.cpp [hunk 4]
```diff
@@ -786,6 +786,43 @@ void SPIRVToLLVM::setLLVMLoopMetadata(const LoopInstType *LM,
...
+  if (LC & LoopControlPipelineEnableINTEL) {
+    Metadata.push_back(llvm::MDNode::get(
+        *Context,
+        getMetadataFromNameAndParameter("llvm.loop.intel.pipelining.enable",
+                                        LoopControlParameters[NumParam++])));
+    assert(NumParam <= LoopControlParameters.size() &&
+           "Missing loop control parameter!");
+  }
+  if (LC & LoopControlLoopCoalesceINTEL) {
+    if (LoopControlParameters.size()) {
```
Comment (AGindinson) [inline lib/SPIRV/SPIRVReader.cpp]:
```suggestion
    if (!LoopControlParameters.empty()) {
```

**Comment by mlychkov** [comment]
LGTM.

**Pull request merged**